### PR TITLE
Check for bad population sizes in SLiM + SLiM logging

### PR DESCRIPTION
**Pull request by grahamgower**

### CODE REVIEW

**Comment by grahamgower** [comment]
This PR add the SLiM engine bits to make all output use the logging module (see #475 for non-SLiM-engine bits). As the first use case, and to avoid silly/trivial unit tests, I've lumped this together with warnings and errors for bad population sizes as raised in #456 and #463.

Example output:
```
$ stdpopsim -e slim --slim-scaling-factor 100 --slim-burn-in 0.1 HomSap -d OutOfAfrica_2T12 50000 50000 --dry-run
Simulation information:
    Engine: slim (3.3.2)
    Model id: OutOfAfrica_2T12
    Model desciption: Two population out-of-Africa
    Population: number_samples (sampling_time_generations):
        AFR: 50000 (0)
        EUR: 50000 (0)
Contig Description:
    Contig length: 249250621.0
    Mean recombination rate: 1.1485597641285933e-08
    Mean mutation rate: 1.29e-08
    Genetic map: None

WARNING: p1 (EUR) has only 18 individuals alive at generation 47.
WARNING: p1 (EUR) has only 10 individuals alive at generation 58.
ERROR: Request to sample 25000 individuals from p0 (AFR) at generation 67, but only 4315 individuals will be alive.
Traceback (most recent call last):
  File "/home/grg/.local/bin/stdpopsim", line 11, in <module>
    load_entry_point('stdpopsim', 'console_scripts', 'stdpopsim')()
  File "/home/grg/src/stdpopsim/stdpopsim/cli.py", line 621, in stdpopsim_main
    run(args)
  File "/home/grg/src/stdpopsim/stdpopsim/cli.py", line 612, in run
    args.runner(args)
  File "/home/grg/src/stdpopsim/stdpopsim/cli.py", line 438, in run_simulation
    ts = engine.simulate(**kwargs)
  File "/home/grg/src/stdpopsim/stdpopsim/slim_engine.py", line 730, in simulate
    raise SLiMException(
stdpopsim.slim_engine.SLiMException: slim exited with code 1.
ERROR (Eidos_ExecuteFunction_stop): stop("Request to sample 25000 individuals from p0 (AFR) at generation 67, but only 4315 individuals will be alive.") called.

Error on script line 3, character 4 (inside runtime script block):

    stop(s);
    ^^^^

```

File: stdpopsim/slim_engine.py [hunk 1]
```diff
@@ -633,16 +704,32 @@ def script_file_f():
...
             script_file.flush()
 
             if not run_slim:
                 return None
 
             slim_cmd.append(script_file.name)
-            stdout = subprocess.DEVNULL if verbosity < 2 else None
-            subprocess.check_call(slim_cmd, stdout=stdout)
+            with subprocess.Popen(
+                    slim_cmd, bufsize=1, universal_newlines=True,
```
Comment (jeromekelleher) [inline stdpopsim/slim_engine.py:716]: I thought on reading this that you had a buffer of size 1 (which sounded like a bad idea), but looking it up you're doing exactly the right thing. Perhaps a quick comment saying "# bufsize=1 means line buffering " or something?

Unusual for Python to be so cryptic!

**Comment by jeromekelleher** [comment]
Great stuff @grahamgower, I really like it. Happy to merge whenever.

We should get someone who understands SLiM to review though. @petrelharp, can you have a look please?

**Comment by grahamgower** [comment]
Oh, and this is the verbose output (all slim output except warnings and errors are converted to DEBUG level output):
```
$ stdpopsim -v -e slim --slim-scaling-factor 10 --slim-burn-in 0.1 HomSap -c chr22 10 -o /dev/null
DEBUG: Making flat chromosome 1 * chr22
DEBUG: Running simulation model PiecewiseConstant for HomSap on Contig(length=5.1E+07, recombination_rate=1.4E-08, mutation_rate=1.3E-08, genetic_map=None) with 10 samples using slim.
Simulation information:
    Engine: slim (3.3.2)
    Model id: PiecewiseConstant
    Model desciption: Piecewise constant size population model over multiple epochs.
    Population: number_samples (sampling_time_generations):
        pop0: 10 (0)
Contig Description:
    Contig length: 51304566.0
    Mean recombination rate: 1.4445022767788226e-08
    Mean mutation rate: 1.29e-08
    Genetic map: None

DEBUG: // Initial random seed:
DEBUG: 10536488367387
DEBUG:
DEBUG: // RunInitializeCallbacks():
DEBUG: initializeTreeSeq();
DEBUG: initializeMutationRate(0);
DEBUG: initializeMutationType(1, 0.5, "f", 0);
DEBUG: initializeGenomicElementType(1, m1, 1);
DEBUG: initializeGenomicElement(g1, 0, 51304565);
DEBUG: initializeRecombinationRate(1.4445e-07, 51304565);
DEBUG:
DEBUG: // Starting run at generation <start>:
DEBUG: 1
DEBUG:
DEBUG: 1: sim.addSubpop(0, 1000);
DEBUG: 1: Starting burn-in...
DEBUG: 101: {inds=p0.sampleIndividuals(5); sim.treeSeqRememberIndividuals(inds);}
DEBUG: 101: {end();}
DEBUG: Loading file version 1.0
DEBUG: Loading 45 items from 14237156 bytes
DEBUG: Closing file '/tmp/tmpwqgq84wd.ts'
DEBUG: rusage: user=17 seconds; sys=0.10s; max_rss=253.9 MiB
DEBUG: Updating provenance
DEBUG: Writing to /dev/null
If you use this simulation in published work, please cite:
[simulation engine]
Haller et al., 2019: https://doi.org/10.1111/1755-0998.12968
[genome assembly]
The Genome Sequencing Consortium, 2001: http://dx.doi.org/10.1038/35057062
[mutation rate]
Tian, Browning, and Browning, 2019: https://doi.org/10.1016/j.ajhg.2019.09.012
[recombination rate]
The International HapMap Consortium, 2007: https://doi.org/10.1038/nature06258
[population size]
Takahata, 1993: https://doi.org/10.1093/oxfordjournals.molbev.a039995
[generation time]
Tremblay and Vezina, 2000: https://doi.org/10.1086/302770
```

File: stdpopsim/slim_engine.py [hunk 2]
```diff
@@ -167,7 +208,8 @@
         for (i in 1:(num_epochs-1)) {
             g = G[i-1];
             for (j in 0:(num_populations-1)) {
-                if (N[i,j] != N[i-1,j]) {
+                if (N[i,j] != N[i-1,j] & N[i-1,j] != 0) {
```
Comment (petrelharp) [inline stdpopsim/slim_engine.py:211]: I don't have the big picture here, but is this right? Can't a population goes from size 0 to something positive?

File: stdpopsim/slim_engine.py [hunk 3]
```diff
@@ -191,6 +233,12 @@
 
                     N0 = N[i,j];
                     r = Q * growth_rates[i,j];
+                    if (r < 0) {
+                        // N decreases, check size at end of "growth" phase.
+                        g_diff = growth_phase_end - growth_phase_start;
+                        N_growth_phase_end = asInteger(round(N0*exp(r*g_diff)));
+                        check_size(j, N_growth_phase_end, growth_phase_end);
```
Comment (petrelharp) [inline stdpopsim/slim_engine.py:240]: Wouldn't it help to factor out a function (or, maybe a lambda) that gives you the size of population `j` at time `t`, say? There's common code below.

File: stdpopsim/slim_engine.py [hunk 4]
```diff
@@ -633,16 +704,42 @@ def script_file_f():
...
 
             if not run_slim:
                 return None
 
+            # Run SLiM.
+            # We capture the output using Popen's line-oriented text buffering
+            # (bufsize=1, universal_newlines=True) and redirect all messages to
+            # Python's logging module.
+            # By convention, messages from SLiM prefixed with "ERROR: " or
+            # "WARNING: " are treated as ERROR or WARN loglevels respectively.
```
Comment (petrelharp) [inline stdpopsim/slim_engine.py:719]: wow, that's neat

File: stdpopsim/slim_engine.py (hunk 2 above)
Comment (grahamgower) [reply stdpopsim/slim_engine.py:211]: It turns out that setting population sizes in this case is redundant (they're all taken care of by either `sim.addSubpop()` or `sim.addSubpopSplit()`). I only just noticed though, due to explicitly counting the warning in the introduced tests.

File: stdpopsim/slim_engine.py (hunk 3 above)
Comment (grahamgower) [reply stdpopsim/slim_engine.py:240]: Hmm, yeah. I factored out the `epoch()` functionality, which is probably the wrong thing. I'll have a go at a pop_size_at() function.

File: stdpopsim/slim_engine.py (hunk 2 above)
Comment (petrelharp) [reply stdpopsim/slim_engine.py:211]: Maybe this warrants an explanatory comment.

**Comment by grahamgower** [comment]
I addressed @petrelharp's comments and made a couple of other small changes.

**Comment by petrelharp** [comment]
Looks good to me! I'm happy to merge if you are (but maybe squash first?).

**Comment by grahamgower** [comment]
Squashed. Thanks!

**Comment by jeromekelleher** [comment]
Since @petrelharp has OK'd I'm going to merge. Thanks @grahamgower, this is super-slick stuff.

**Pull request merged**